x-pack/plugin/repository-encrypted/src/main/java/org/elasticsearch/repositories/encrypted/DecryptionPacketsInputStream.java

```diff
@@ -0,0 +1,158 @@
+    }
+
+    public DecryptionPacketsInputStream(InputStream source, SecretKey secretKey, int nonce, int packetLength) {
+        this.source = Objects.requireNonNull(source);
+        this.secretKey = Objects.requireNonNull(secretKey);
+        this.nonce = nonce;
+        if (packetLength <= 0 || packetLength >= EncryptedRepository.MAX_PACKET_LENGTH_IN_BYTES) {
+            throw new IllegalArgumentException("Invalid packet length [" + packetLength + "]");
+        }
+        this.packetLength = packetLength;
+        this.packet = new byte[packetLength + EncryptedRepository.GCM_TAG_LENGTH_IN_BYTES];
+        this.iv = new byte[EncryptedRepository.GCM_IV_LENGTH_IN_BYTES];
+        this.hasNext = true;
+        this.counter = EncryptedRepository.PACKET_START_COUNTER;
+    }
+
+    @Override
+    InputStream nextComponent(InputStream currentComponentIn) throws IOException {
+        if (currentComponentIn != null && currentComponentIn.read() != -1) {
+            throw new IllegalStateException("Stream for previous packet has not been fully processed");
+        }
+        if (false == hasNext) {
+            return null;
+        }
+        PrefixInputStream packetInputStream = new PrefixInputStream(source,
+                packetLength + EncryptedRepository.GCM_IV_LENGTH_IN_BYTES + EncryptedRepository.GCM_TAG_LENGTH_IN_BYTES,
+                false);
+        int currentPacketLength = decrypt(packetInputStream);
+        // only the last packet is shorter, so this must be the last packet
+        if (currentPacketLength != packetLength) {
+            hasNext = false;
+        }
+        return new ByteArrayInputStream(packet, 0, currentPacketLength);
+    }
+
+    @Override
+    public boolean markSupported() {
+        return false;
+    }
+
+    @Override
+    public void mark(int readlimit) {
+    }
+
+    @Override
+    public void reset() throws IOException {
+        throw new IOException("Mark/reset not supported");
+    }
+
+    private int decrypt(PrefixInputStream packetInputStream) throws IOException {
+        if (packetInputStream.read(iv) != iv.length) {
+            throw new IOException("Error while reading the heading IV of the packet");
+        }
+        ByteBuffer ivBuffer = ByteBuffer.wrap(iv).order(ByteOrder.LITTLE_ENDIAN);
+        if (ivBuffer.getInt(0) != nonce || ivBuffer.getLong(4) != counter++) {
+            throw new IOException("Invalid packet IV");
+        }
+        int packetLength = packetInputStream.read(packet);
+        if (packetLength < EncryptedRepository.GCM_TAG_LENGTH_IN_BYTES) {
+            throw new IOException("Error while reading the packet");
+        }
+        Cipher packetCipher = getPacketDecryptionCipher(iv);
+        try {
+            // in-place decryption
+            return packetCipher.doFinal(packet, 0, packetLength, packet);
+        } catch (ShortBufferException | IllegalBlockSizeException | BadPaddingException e) {
+            throw new IOException(e);
```

Comment:
Can we add a message to this exception?

Reply:
Sure, I've pushed fd109144d88a98ea4ca05 I don't think there's much context information available anyway.